lib/src/supabase_auth.dart

```diff
@@ -96,6 +106,48 @@ class SupabaseAuth {
   void dispose() {
     _listenerController.close();
     _authSubscription?.data?.unsubscribe();
+    _stopDeeplinkObserver();
+    WidgetsBinding.instance?.removeObserver(this);
+  }
+
+  @override
+  void didChangeAppLifecycleState(AppLifecycleState state) {
+    switch (state) {
+      case AppLifecycleState.resumed:
+        _recoverSupabaseSession();
+        break;
+      case AppLifecycleState.inactive:
+        break;
+      case AppLifecycleState.paused:
+        break;
+      case AppLifecycleState.detached:
+        break;
+    }
+  }
+
+  /// Recover/refresh session if it's available
+  /// e.g. called on a Splash screen when app starts.
+  Future<bool> _recoverSupabaseSession() async {
+    final bool exist =
+        await SupabaseAuth.instance.localStorage.hasAccessToken();
+    if (!exist) {
+      return false;
+    }
+
+    final String? jsonStr =
+        await SupabaseAuth.instance.localStorage.accessToken();
+    if (jsonStr == null) {
+      return false;
+    }
+
+    final response =
+        await Supabase.instance.client.auth.recoverSession(jsonStr);
```

Comment:
Do we want to recover a session, (Which I think is refreshing the accessToken), on every state resume? Or only if the user is currently not logged in / if the access token is expired?

Reply:
>I'd want _instance.isAuthCallbackDeeplink to be better at checking to see if it's actually a url that we want to recover a session from.

This is a great point! Will edit the logic to a more robust way of checking. 

>Do we want to recover a session, (Which I think is refreshing the accessToken), on every state resume? Or only if the user is currently not logged in / if the access token is expired?

For this one, I think we do want to recover session on every state resume (every time the app comes back to foreground from background), because the timer set to recover session stops when the app is in background, so we want to get a new token every time we come back to foreground, but do you think this can be optimized?

Reply:
About the session recovery on every time app coming back to foreground, I believe the js sdk has something similar implemented, and generally token refresh is not an expensive thing to do, so it's okay to over do it!

Reply:
@DanMossa 
>I'd want _instance.isAuthCallbackDeeplink to be better at checking to see if it's actually a url that we want to recover a session from.

Did you have any particular idea of what this could look like? I was going to implement a check to see if `#` is included in the URL or something, but started thinking that might not be a very good idea if the Gotrue server decided not to use `#` in the future. 

Also, even if we pass deeplinks that are actually not Supabase auth related, it will be rejected [here](https://github.com/supabase-community/gotrue-dart/blob/main/lib/src/gotrue_client.dart#L230), so right now, I am leaning towards leaving the way it is, but what do you think?